src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -183,12 +209,54 @@ public SocketAsyncContext(int fileDescriptor, SocketAsyncEngine engine)
                 _handle.Free();
 
                 // TODO: throw an appropiate exception
-                throw new Exception(string.Format("SocketAsyncContext: {0}", errorCode));
+                throw new Exception(string.Format("SocketAsyncContext.Register: {0}", errorCode));
+            }
+
+            _registeredEvents = events;
+        }
+
+        private void UnregisterRead()
+        {
+            Debug.Assert(Monitor.IsEntered(_queueLock));
+            Debug.Assert(_registeredEvents == (SocketAsyncEvents.Read | SocketAsyncEvents.Write));
+
+            SocketAsyncEvents events = _registeredEvents & ~SocketAsyncEvents.Read;
+
+            Interop.Error errorCode;
+            if (!_engine.TryRegister(_fileDescriptor, _registeredEvents, events, _handle, out errorCode))
+            {
+                // TODO: throw an appropiate exception
+                throw new Exception(string.Format("UnregisterRead: {0}", errorCode));
             }
 
             _registeredEvents = events;
         }
 
+        private void Unregister()
+        {
+            Debug.Assert(Monitor.IsEntered(_queueLock));
+
+            if (_registeredEvents == SocketAsyncEvents.None)
+            {
+                Debug.Assert(!_handle.IsAllocated);
+                return;
+            }
+
+
+            Interop.Error errorCode;
+            if (!_engine.TryRegister(_fileDescriptor, _registeredEvents, SocketAsyncEvents.None, _handle, out errorCode))
+            {
+                if (errorCode != Interop.Error.EBADF)
+                {
+                    // TODO: throw an appropiate exception
+                    throw new Exception(string.Format("Unregister: {0}", errorCode));
+                }
+            }
+
+            _handle.Free();
```

Comment:
Yes, I think so. We also need some way to indicate that unregistration failed for a particular socket, so if events are received on that socket, the handler should retry the unregistration.